File: GameServer/GameServerApp/GameServerApp/Proto/CreateLuaProto/RoleSkillDataResponseProto.lua
--角色技能主动通知协议
RoleSkillDataResponseProto = { ProtoCode = 31001, SkillCount = 0, CurrSkillDataTable = { } }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
RoleSkillDataResponseProto.__index = RoleSkillDataResponseProto;

function RoleSkillDataResponseProto.New()
    local self = { }; --初始化self
    setmetatable(self, RoleSkillDataResponseProto); --将self的元表设定为Class
    return self;
end


--定义当前学会的技能
CurrSkillData = { SkillId = 0, SkillLevel = 0, SlotsNode = 0 }
CurrSkillData.__index = CurrSkillData;
function CurrSkillData.New()
    local self = { };
    setmetatable(self, CurrSkillData);
    return self;
end


--发送协议
function RoleSkillDataResponseProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteByte(proto.SkillCount);
    for i = 1, proto.SkillCount, 1 do
        ms:WriteInt(CurrSkillDataList[i].SkillId);
        ms:WriteInt(CurrSkillDataList[i].SkillLevel);
        ms:WriteByte(CurrSkillDataList[i].SlotsNode);
    end

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function RoleSkillDataResponseProto.GetProto(buffer)

    local proto = RoleSkillDataResponseProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.SkillCount = ms:ReadByte();
	proto.CurrSkillDataTable = {};
    for i = 1, proto.SkillCount, 1 do
        local _CurrSkillData = CurrSkillData.New();
        _CurrSkillData.SkillId = ms:ReadInt();
        _CurrSkillData.SkillLevel = ms:ReadInt();
        _CurrSkillData.SlotsNode = ms:ReadByte();
        proto.CurrSkillDataTable[#proto.CurrSkillDataTable+1] = _CurrSkillData;
    end

    ms:Dispose();
    return proto;
end

File: MMORPG_2/Assets/Script/Data/Proto/CreateProto/RoleSkillDataResponseProto.cs
//===================================================
//作    者：xxx
//创建时间：2019-05-18 15:13:21
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary> 角色技能主动通知协议 </summary>
public struct RoleSkillDataResponseProto : IProto
{
    public ushort ProtoCode { get { return 31001; } }

    /// <summary> 学会的技能数量 </summary>
    public byte SkillCount;
    /// <summary> 当前学会的技能 </summary>
    public List<SkillData> CurrSkillDataList;

    /// <summary> 当前学会的技能 </summary>
    public struct SkillData
    {
        /// <summary> 技能编号 </summary>
        public int SkillId;
        /// <summary> 技能等级 </summary>
        public int SkillLevel;
        /// <summary> 技能槽编号 </summary>
        public byte SlotsNode;
    }

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteByte(SkillCount);
            for (int i = 0; i < SkillCount; i++)
            {
                ms.WriteInt(CurrSkillDataList[i].SkillId);
                ms.WriteInt(CurrSkillDataList[i].SkillLevel);
                ms.WriteByte(CurrSkillDataList[i].SlotsNode);
            }
            return ms.ToArray();
        }
    }

    public static RoleSkillDataResponseProto GetProto(byte[] buffer)
    {
        RoleSkillDataResponseProto proto = new RoleSkillDataResponseProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.SkillCount = (byte)ms.ReadByte();
            proto.CurrSkillDataList = new List<SkillData>();
            for (int i = 0; i < proto.SkillCount; i++)
            {
                SkillData _CurrSkillData = new SkillData();
                _CurrSkillData.SkillId = ms.ReadInt();
                _CurrSkillData.SkillLevel = ms.ReadInt();
                _CurrSkillData.SlotsNode = (byte)ms.ReadByte();
                proto.CurrSkillDataList.Add(_CurrSkillData);
            }
        }
        return proto;
    }
}